File: ruby/tournament/tournament.rb
class Team
  attr_accessor :name, :wins, :draws, :losses

  def initialize(name)
    @name = name
    @wins = 0
    @draws = 0
    @losses = 0
  end

  def points
    wins * 3 + draws
  end

  def played
    wins + draws + losses
  end
end

class TournamentTable
  attr_accessor :teams

  def self.render(teams)
    new(teams).render
  end

  def initialize(teams)
    @teams = teams
  end

  def render
    header + rows
  end

  private

  def header
    "Team                           | MP |  W |  D |  L |  P\n"
  end

  def rows
    teams.map(&method(:row)).join
  end

  def row(team)
    format("%-30s | %2d | %2d | %2d | %2d | %2d\n",
           team.name, team.played, team.wins, team.draws, team.losses, team.points)
  end
end

class Tournament
  attr_reader :teams

  def self.tally(results)
    new(results).tally
  end

  def initialize(results)
    @teams = to_teams(results.split("\n"))
  end

  def tally
    TournamentTable.render(teams)
  end

  private

  def to_teams(matches)
    matches
      .each_with_object({}, &method(:process_match))
      .values
      .sort_by { |team| [-team.points, team.name] }
  end

  def process_match(match, teams)
    team1, team2, result = match.split(';')

    teams[team1] ||= Team.new(team1)
    teams[team2] ||= Team.new(team2)

    case result
    when 'loss'
      teams[team1].losses += 1
      teams[team2].wins += 1
    when 'win'
      teams[team1].wins += 1
      teams[team2].losses += 1
    when 'draw'
      teams[team1].draws += 1
      teams[team2].draws += 1
    end
  end
end

File: typescript/tournament/tournament.ts
class Team {
  public wins = 0;
  public draws = 0;
  public losses = 0;

  constructor(public name: string) {}

  get played(): number {
    return this.wins + this.draws + this.losses;
  }

  get points(): number {
    return this.wins * 3 + this.draws;
  }
}

class Report {
  private lines: string[] = [];

  constructor() {
    this.addRow("Team", "MP", "W", "D", "L", "P");
  }

  public addRow(
    name: string,
    played: string,
    wins: string,
    draws: string,
    losses: string,
    points: string
  ): void {
    this.lines.push(
      [
        name.padEnd(30),
        played.padStart(2),
        wins.padStart(2),
        draws.padStart(2),
        losses.padStart(2),
        points.padStart(2),
      ].join(" | ")
    );
  }

  public toString(): string {
    return this.lines.join("\n");
  }
}

export class Tournament {
  private teams: { [name: string]: Team } = {};

  public tally(input: string): string {
    this.importScores(input);

    const report = new Report();

    for (const team of Object.values(this.teams).sort(
      (a, b) => b.points - a.points || a.name.localeCompare(b.name)
    )) {
      report.addRow(
        team.name,
        team.played.toString(),
        team.wins.toString(),
        team.draws.toString(),
        team.losses.toString(),
        team.points.toString()
      );
    }

    return report.toString();
  }

  private importScores(input: string): void {
    for (const line of input.split("\n").filter((s) => s.length > 0)) {
      const [home, away, result] = line.split(";");

      this.teams[home] ||= new Team(home);
      this.teams[away] ||= new Team(away);

      switch (result) {
        case "win":
          this.teams[home].wins++;
          this.teams[away].losses++;
          break;
        case "draw":
          this.teams[home].draws++;
          this.teams[away].draws++;
          break;
        case "loss":
          this.teams[home].losses++;
          this.teams[away].wins++;
          break;
      }
    }
  }
}
